docs/development.rst

```diff
@@ -2,9 +2,208 @@
     :start-after: sec-begin-development
     :end-before: sec-end-development
 
+.. _writing-adapters:
+
 Writing a model adapter
 =======================
 
+Creating an :class:`~openscm.adapter.Adapter` subclass
+******************************************************
+
+- Create your adapter source file in ``openscm/adapters/``, e.g.
+  ``MyAdapter.py``
+
+- Do not add additional dependencies to OpenSCM (such as in
+  ``setup.py``), but make sure that you throw an
+  :class:`~openscm.errors.AdapterNeedsModuleError` when a module you
+  depend on is not installed on the user's system (please include a
+  message how to install the missing module):
+
+  .. code:: python
+
+      from ..errors import AdapterNeedsModuleError
+
+      try:
+          import MyExternalModule
+      except ImportError:
+          raise AdapterNeedsModuleError(
+              """
+              To run MyModel the 'MyExternalModule' is needed.
+              Please install it using `pip install myexternalmodule`.
+              """
+          )
+
+- Subclass the :class:`openscm.adapter.Adapter` class:
+
+  .. code:: python
+
+      from ..adapter import Adapter
+
+      class MyAdapter(Adapter):
+
+- Implement the relevant methods (you are not obliged to implement any
+  method, but should at least implement the
+  :func:`~openscm.adapter.Adapter.run` function to do the actual model
+  run. An adapter should ONLY interface with the :ref:`low-level API
+  <low-level-reference>`.
+
+  Make sure to always call the original implementation of each
+  implemented method using ``super.FUNCTION_NAME(PARAMETERS)`` (see
+  examples below). The implemented functions should NOT accept
+  additional parameters (e.g. using ``*args``/``**kwargs``).
+
+  - The :func:`~openscm.adapter.Adapter.__init__` method initializes
+    the adapter. It should set the default values of mandatory
+    *model-specific* (not :ref:`standard OpenSCM parameters
+    <standard-parameters>`!) parameters in the
+    :class:`~openscm.core.ParameterSet`. Their :ref:`hierarchical
+    names <parameter-hierarchy>` should start with the model/adapter
+    name.
+
+    .. code:: python
+
+        def __init__(self, parameters: ParameterSet):
+            super.__init__(parameters)
+            # TODO Initialize the model
+            # TODO Set default parameter values:
+            self._parameters.get_writable_scalar_view(
+                ("MyModel", "Specific Parameter"), ("World",), "Unit"
+            ).set(DEFAULT_VALUE)
+
+  - The :func:`~openscm.adapter.Adapter.initialize_run_parameters`
```

Comment:
1. Do we need an initialization function for the model itself apart from ``__init__``? Though we could use that for easier implementation without having to call the ``super.XX`` methods... Will think about that. Maybe having a separate ``shutdown`` method (apart from ``__del__``) would be more explicit?
2./3//4. I would prefer to have the same input ParameterSet (containing both input values and time series as well as model parameters) for different runs (that way ParameterViews can be reused and have only to be eestablished by ``__init__`` but read from in the other initialization functions). So, I would not have them as parameters here. Start and stop time would also be easier to know here already to properly handle the retreived time series, I suppose. Also, for stepwise running we need to know ``start_time`` and ``stop_time`` beforehand.

(Question aside: naming in american or british english?)

Reply:
@znicholls @rgieseke ?

Reply:
>     * Do we need an initialization function for the model itself apart from `__init__`?

I don't think so. I've found it useful whilst playing around as I can then separate `adapter = MAGICC6()` and `adapter.initialise(root_dir="somewhere")` but I'm not sure it's actually a necessary separation long term.

>     * Maybe having a separate `shutdown` method (apart from `__del__`) would be more explicit?

I'm happy with whatever you decide here. Something to automatically mop up after a run would be nice (mainly for models which write a bunch of input/output to disc like MAGICC and HECTOR (?)).

> naming in american or british english?

Preference for british but don't mind

>     * I would prefer to have the same input ParameterSet (containing both input values and time series as well as model parameters) for different runs (that way ParameterViews can be reused and have only to be eestablished by `__init__` but read from in the other initialization functions). So, I would not have them as parameters here.

To check I've understood. On `__init__`, the adapter would return a `Core` with all the views established that the adapter understands. The user (coming from whatever level) uses the pre-initialised views to set start time and stop time and parameters and hence gets errors if they try to set anything which the adapter doesn't understand. Then the model just runs from `Core` (as it stands when `run` is called), returning a different `Core` instance with the results for that given run. If yes, that sounds genius.

Reply:
Ok, I have now moved all methods for the user to implement to abstract methods, so there are ``_initialize_model`` and ``_shutdown`` now.

> To check I've understood. On __init__, the adapter would return a Core with all the views established that the adapter understands.

The (now) ``_initialize_model`` method only fills the input parameter set. The ``Adapter`` class is only called from the ``Core`` class, but does not have anything to do with ``Core`` apart from that. The ``ParameterSet`` objects stay in the adapter. The adapter should not use views for non-model-specific parameters yet (I have to document that).

> The user (coming from whatever level) uses the pre-initialised views to set start time and stop time and parameters and hence gets errors if they try to set anything which the adapter doesn't understand.

The user uses the ``ParameterSet`` object (not the views, but these refer to that object) for all the parameters. They do not receive errors when setting parameters, only when the adapter pulls them, but for certain parameters expects them in a way that is impossible to convert. Parameters that the adapter doesn't understand (i.e. know about) would only remain unused (would implement a list of unused parameters the user can get).

> Then the model just runs from Core (as it stands when run is called), returning a different Core instance with the results for that given run. If yes, that sounds genius (I'm guessing the time penalty of creating views is zero as they're just views).

When the adapter is run (being called from ``Core``) it reads the new values from the (pre-instantiated) views and updates the output parameters via views (again). The user can use their views from the output parameter set to receive the (wanted) output.

Reply:
> The `ParameterSet` objects stay in the adapter.

What does this mean? It's not passed back up to the user to then be able to fill out?

> The user uses the `ParameterSet` object

The one that stays in the adapter?

> only when the adapter pulls them

So every run, the adapter pulls every single parameter it knows about? Will this incur to large a processing cost if we only want to change one value?

> The user can use their views

Where are these views collated/collected/stored?

Reply:
> > The `ParameterSet` objects stay in the adapter.
> 
> What does this mean? It's not passed back up to the user to then be able to fill out?

Its main reference is ``Adapter._parameters`` so it kind of belongs to an adapter, but is passed to the end-user from the corresponding ``Core`` instance.

> > The user uses the `ParameterSet` object
> 
> The one that stays in the adapter?

Yes (see above)

> > only when the adapter pulls them
> 
> So every run, the adapter pulls every single parameter it knows about? Will this incur to large a processing cost if we only want to change one value?

If it keeps the view that should not be too much of an overhead, but could implement a ``_Parameter.version`` to check for new data. Alternatively, we could use callback functions for each view that are called each time the user changes data (though that seems a bit much, especially when you have a model that does bot cache the input parameters and where the adapter needs to get all input data again after a run, anyway....).

> > The user can use their views
> 
> Where are these views collated/collected/stored?

Up to the end-user ;) In the ``OpenSCM`` instance this would probably done in the background (have an idea about that interface that I will push to #118 in a week.